### PR TITLE
Fix spin-up missing deeper subfolders. Fix AWS API getting overloaded during tests. 

### LINKED ISSUE #81: Spin-up misses modules in subfolders.
Given the following folder structure:

```
infrastructure-live
- stage
  - data-stores
    - postgres
    - redis
  - ecs-cluster
  - kms-master-key
  - vpc
```

I ran `terragrunt spin-up` and expected Terragrunt to apply all the above folders as long as each had a `.terragrunt` file. But the `stage/data-stores` is ignored, even though both `stage/data-stores/postgres` and `stage/data-stores/redis` are confirmed to have a `.terragrunt` file .

Here's a [zip file of the full infrastructure-live directory](https://drive.google.com/open?id=0B161Wobnisu-cVo0RHE3bzJESVk) in a private Google Drive folder available to Gruntworkers only so someone else can reproduce.

Here's the output. Note the inclusion of ignored modules.

```
josh@Joshs-MacBook-Pro /r/g/b/i/stage> terragrunt spin-up
[terragrunt]  Module /repos/gruntwork-clients/acme/infrastructure-live/stage/vpc depends on module /repos/gruntwork-clients/acme/infrastructure-live/mgmt/vpc, which is an external dependency outside of the current working directory. Should Terragrunt skip over this external dependency? Warning, if you say 'no', Terragrunt will make changes in /repos/gruntwork-clients/acme/infrastructure-live/mgmt/vpc as well! (y/n) y
[terragrunt]  Module /repos/gruntwork-clients/acme/infrastructure-live/stage/ecs-cluster depends on module /repos/gruntwork-clients/acme/infrastructure-live/mgmt/bastion-host, which is an external dependency outside of the current working directory. Should Terragrunt skip over this external dependency? Warning, if you say 'no', Terragrunt will make changes in /repos/gruntwork-clients/acme/infrastructure-live/mgmt/bastion-host as well! (y/n) y
[terragrunt]  Module /repos/gruntwork-clients/acme/infrastructure-live/stage/ecs-cluster depends on module /repos/gruntwork-clients/acme/infrastructure-live/global/sns-topics, which is an external dependency outside of the current working directory. Should Terragrunt skip over this external dependency? Warning, if you say 'no', Terragrunt will make changes in /repos/gruntwork-clients/acme/infrastructure-live/global/sns-topics as well! (y/n) y
[terragrunt] 2016/12/19 15:07:23 Stack at /repos/gruntwork-clients/acme/infrastructure-live/stage:
  => Module /repos/gruntwork-clients/acme/infrastructure-live/stage/ecs-cluster (dependencies: [/repos/gruntwork-clients/acme/infrastructure-live/stage/vpc, /repos/gruntwork-clients/acme/infrastructure-live/mgmt/bastion-host, /repos/gruntwork-clients/acme/infrastructure-live/global/sns-topics])
  => Module /repos/gruntwork-clients/acme/infrastructure-live/stage/kms-master-key (dependencies: [])
  => Module /repos/gruntwork-clients/acme/infrastructure-live/stage/vpc (dependencies: [/repos/gruntwork-clients/acme/infrastructure-live/mgmt/vpc])
  => Module /repos/gruntwork-clients/acme/infrastructure-live/mgmt/vpc (dependencies: [])
  => Module /repos/gruntwork-clients/acme/infrastructure-live/mgmt/bastion-host (dependencies: [/repos/gruntwork-clients/acme/infrastructure-live/mgmt/vpc, /repos/gruntwork-clients/acme/infrastructure-live/global/sns-topics])
  => Module /repos/gruntwork-clients/acme/infrastructure-live/global/sns-topics (dependencies: [])
[terragrunt]  Are you sure you want to run 'terragrunt apply' in each folder of the stack described above? (y/n)
```

### PR DESCRIPTION
**Pull request by josh-padnick**
This PR fixes two separate issues:

- First, it allows `spin-up` and `tear-down` to look for a `.terragrunt` file in more folders than just the immediate subfolders. I implemented unit tests to validate this functionality.
- Second, Terragrunt tests were failing arbitrarily when the AWS API complained of being too much at once. This was because when many parallel Terragrunt tests created a DynamoDB Table and then waited 10 seconds to check again, all checks came at exactly the same time. I resolved this by introducing a random sleep between tests while keeping the API the same for non-test code (i.e. no random sleep for non-test code).

Fixes #81.

### CODE REVIEW

File: spin/stack.go [hunk 1]
```diff
@@ -47,7 +49,10 @@ func (stack *Stack) CheckForCycles() error {
 // Find all the Terraform modules in the subfolders of the working directory of the given TerragruntOptions and
 // assemble them into a Stack object that can be applied or destroyed in a single command
 func FindStackInSubfolders(terragruntOptions *options.TerragruntOptions) (*Stack, error) {
-	terragruntConfigFiles, err := filepath.Glob(fmt.Sprintf("%s/**/%s", terragruntOptions.WorkingDir, config.DefaultTerragruntConfigPath))
+	// Ideally, we'd use a builin Go library like filepath.Glob here, but per https://github.com/golang/go/issues/11862,
+	// the current go implementation doesn't support treating ** as zero or more directories, just zero or one.
+	// So we use a third-party library.
```
Comment (brikis98) [inline spin/stack.go]: Good catch.

I really think Go is a shitty language. Don't name the method `Glob` if it doesn't implementing globbing the way 99.9% of platforms do. And then refuse to fix clearly broken behavior in the name of backwards compatibility.

File: util/random.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+func GetRandomTime(lowerBound, upperBound time.Duration) time.Duration {
+	if lowerBound == upperBound {
+		return lowerBound
+	}
+
+	var lowerBoundMs, upperBoundMs float64
+
+	if upperBound < 1 * time.Second {
+		lowerBoundMs = 1
+		upperBoundMs = 1000
```
Comment (brikis98) [inline util/random.go]: Why this special case?

File: util/random.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"time"
+	"math/rand"
+)
+
+// Get a random time duration between the lower bound and upper bound. This is useful because some of our automated tests
+// wound up flooding the AWS API all at once, leading to a "Subscriber limit exceeded" error.
+func GetRandomTime(lowerBound, upperBound time.Duration) time.Duration {
+	if lowerBound == upperBound {
+		return lowerBound
+	}
```
Comment (brikis98) [inline util/random.go]: What about the case where `lowerBound < upperBound`?

File: util/random_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+func TestGetRandomTime(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct{
+		lowerBound time.Duration
+		upperBound time.Duration
+	}{
+		{ 1 * time.Second, 10 * time.Second },
```
Comment (brikis98) [inline util/random_test.go]: Might be worth testing a bunch of different ranges here, such as:

* (0, 0)
* (-1, -3)
* (1, 1.000001)
* (1, 10)

It may also be worth running a loop for each test case and testing it multiple times, as there are random numbers involved, so it's possible a test will pass as a pure fluke simply because the random number didn't catch a corner case.

File: glide.lock [hunk 5]
```diff
@@ -1,8 +1,8 @@
 hash: f70d8455d9f5f68d4449263ab8d0485d5b938b65b4c31716be50aac64fedc6ae
```
Comment (brikis98) [inline glide.lock]: I think you're missing `glide.yaml` in this PR.

File: spin/stack.go (hunk 1 above)
Comment (josh-padnick) [inline spin/stack.go]: Yeah, I hear you. Their minimalism is great for the learning curve but you constantly find yourself re-implementing boilerplate.

File: util/random.go (hunk 3 above)
Comment (josh-padnick) [inline util/random.go]: Ah, that's a good point, but what to return? I'll play around with a couple of options.

File: util/random.go (hunk 2 above)
Comment (josh-padnick) [inline util/random.go]: It was an early implementation vestige. The thinking was that at a certain level, time ranges are so small as to not be meaningful, but no reason I need to enforce that. Will remove.

File: util/random.go (hunk 3 above)
Comment (brikis98) [inline util/random.go]: Either an error or `upperBound`, I guess.

File: glide.lock (hunk 5 above)
Comment (josh-padnick) [inline glide.lock]: Well, tests pass, so I guess `glide.lock` is good enough, but I manually added to `glide.yaml` for consistency's sake.

Comment (brikis98) [inline glide.lock]: Without `glide.yaml`, the next time someone runs `glide up`, it'll lose the new zglob dependency.

**Comment by josh-padnick** [comment]
Ok, all comments responded to. If this passes tests, unless you strongly object to some of the edge cases I left out of `util.GetRandomTime()`, I think we're good to go here.

File: util/random.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+		upperBound = -1 * upperBound
+	}
+
+	if lowerBound > upperBound {
+		return upperBound
+	}
+
+	if lowerBound == upperBound {
+		return lowerBound
+	}
```
Comment (brikis98) [inline util/random.go]: You could remove this case and just change the previous one to `>=` for exactly the same behavior.

**Comment by brikis98** [comment]
(Feel free to merge when ready)
